pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotTableConfigRestletResource.java

```diff
@@ -0,0 +1,188 @@
+import com.linkedin.pinot.common.config.Deserializer;
+import com.linkedin.pinot.common.config.Serializer;
+import com.linkedin.pinot.common.config.TableConfig;
+import com.linkedin.pinot.common.data.Schema;
+import com.linkedin.pinot.common.utils.CommonConstants;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import io.swagger.annotations.ApiOperation;
+import java.io.IOException;
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Path("/")
+public class PinotTableConfigRestletResource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotTableConfigRestletResource.class);
+
+  @Inject
+  private PinotHelixResourceManager _resourceManager;
+
+  @GET
+  @Produces("application/hocon")
+  @Path("/v2/tables/{tableName}")
+  @ApiOperation("Displays the configuration of a table")
+  public Response readTableConfiguration(
+      @PathParam("tableName") String tableName
+  ) {
+    TableConfig offlineTableConfig =
+        _resourceManager.getTableConfig(tableName, CommonConstants.Helix.TableType.OFFLINE);
+    TableConfig realtimeTableConfig =
+        _resourceManager.getTableConfig(tableName, CommonConstants.Helix.TableType.REALTIME);
+    Schema tableSchema = _resourceManager.getTableSchema(tableName);
+
+    if (offlineTableConfig == null && realtimeTableConfig == null) {
+      return Response
+          .status(Response.Status.NOT_FOUND)
+          .build();
+    }
+
+    CombinedConfig combinedConfig = new CombinedConfig(offlineTableConfig, realtimeTableConfig, tableSchema);
+    String serializedConfig = Serializer.serializeToString(combinedConfig);
+
+    return Response
+        .ok(serializedConfig)
+        .header("Content-Disposition", "inline")
+        .build();
+  }
+
+  @POST
+  @Produces(MediaType.APPLICATION_JSON)
+  @Path("/v2/tables")
+  public Response createNewTable(String tableConfiguration) {
+    CombinedConfig config = null;
+
+    try {
+      config = Deserializer.deserializeFromString(CombinedConfig.class, tableConfiguration);
+    } catch (Exception e) {
+      LOGGER.warn("Caught exception while deserializing the table configuration", e);
+      return Response
+          .serverError()
+          .entity(e.getMessage())
+          .type(MediaType.TEXT_PLAIN_TYPE)
+          .build();
+    }
+
+    if (config == null) {
+      LOGGER.warn("Failed to deserialize the table configuration: {}", tableConfiguration);
+      return Response
+          .serverError()
+          .entity("Failed to deserialize the table configuration")
+          .type(MediaType.TEXT_PLAIN_TYPE)
+          .build();
+    }
+
+    if (config.getSchema() != null) {
+      _resourceManager.addOrUpdateSchema(config.getSchema());
```

Comment:
It looks like resourceManager APIs can throw runtime exceptions some of which indicate client side errors (ie, broker tenant does not exist etc). Shouldn't we catch those and throw appropriate 4xxs?

(There are more occurences of this in the rest of the file)

Reply:
None of the 400 series of error codes are appropriate for this purpose. If there is an exception that is thrown by the restlet, this results in 500 (just like other error handlers), which is the expected code.

Reply:
I am not sure if I follow. If someone tries to add a new table with a broker tenant that does not exist, is that not a client error? In which case, isn't 400 the right error code?

Reply:
The request is fully formed and syntactically correct, so HTTP status code 400 does not apply in this case.

Reply:
Since this has been blocked for over two months, what is the expected code change here. It seems that this pull request is blocked over a minor squabble over which HTTP status code should be returned instead of having a blocker on the design or anything major.

Considering this API isn't in active use and can be changed later, that this API not being available blocks adding the appropriate client tools for the new configuration format so that we can actually exercise the feature in order to make sure that it does work as expected, it seems like this level of scrutiny misses the forest for the trees at this point.